[Quant] Add fused ConvAdd2d module for onednn backend

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91155
* #91154
* #91153
* __->__ #91152

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
[Quant] Add fused ConvAdd2d module for onednn backend

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] [Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] [Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] [Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] [Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP] [Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 reviewed (approved):
LGTM.

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -472,17 +508,17 @@ def test_conv1d_api(self):
             ["zeros", "reflect"],  # pad_mode
             [True, False],  # use_bias
-            [True, False],  # use_fused
+            ["none", "relu"],  # post_op
```
nit: can be done in a separate PR, but we can split this into multiple tests as well

leslie-fang-intel replied:
Thanks for the suggestions. Split this into multiple tests and so for the other similar test cases.

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -537,16 +573,16 @@ def test_conv2d_api(self):
         options = itertools.product(
             [True, False],  # use_bias
-            [True, False],  # use_fused
+            ["relu", "none"],  # post_op
```
same here

leslie-fang-intel replied:
Thanks for the suggestions. Split this into multiple tests.

jerryzh168 commented on torch/ao/nn/intrinsic/quantized/modules/conv_add.py:
```diff
@@ -0,0 +1,50 @@
+
+    @classmethod
+    def from_float(cls, mod):
+        return super(ConvAdd2d, cls).from_float(mod)
```
nit: `super(ConvAdd2d, cls)` --> `super()`?

jerryzh168 commented on torch/ao/nn/intrinsic/quantized/modules/conv_add.py:
```diff
@@ -0,0 +1,50 @@
+    def __init__(self, in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1, bias=True,
+                 padding_mode='zeros', device=None, dtype=None):
+        super(ConvAdd2d, self).__init__(
```
nit: `super(ConvAdd2d, self)` --> `super()`?

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on torch/ao/nn/intrinsic/quantized/modules/conv_add.py:
```diff
@@ -0,0 +1,50 @@
+
+    @classmethod
+    def from_float(cls, mod):
+        return super(ConvAdd2d, cls).from_float(mod)
```
not sure if `super()` works here as well, but maybe can try that in the following PRs that actually tests this logic

leslie-fang-intel replied:
Thanks for the suggestions. According to the pep-3135: https://peps.python.org/pep-3135/. It seems works and I changed it accordingly.

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -274,14 +274,26 @@ def _test_conv_api_impl(
             batch_size, in_channels_per_group, input_feature_map_size,
             out_channels_per_group, groups, kernel_size, X_scale, X_zero_point,
             W_scale, W_zero_point, use_bias, use_channelwise)
+
```
there are a lot of if/else here in this function, I'm wondering if we can pass them as arguments,
e.g. can we pass 
1). inputs to the module as arguments ([[X_q, X2_q]] if (post_op == "add") else [[X_q]])
2). separate the fused_conv_module and the individual conv_module as two arguments, to avoid branches in L296-303
3). any other refactors to remove branches

leslie-fang-intel replied:
Thanks for the suggestions. I have removed most of the if/else branches in this function. Please help to take a look again @jerryzh168.

jerryzh168 reviewed (changes requested):
I think we should remove the branches as much as we can to make the code clearer

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
> I think we should remove the branches as much as we can to make the code clearer

Most of the branches has been removed.

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, thanks for your review. the previous comments has been fixed. Could you kindly take a look of this PR again?

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -512,13 +567,64 @@ def test_conv2d_api(self):
+        for pad_mode, use_bias, use_channelwise in options:
+            if torch.backends.quantized.engine == "qnnpack":
+                use_channelwise = False
+            batch_size = 2
+            in_channels_per_group = 2
+            H = 8
+            W = 8
+            out_channels_per_group = 2
+            groups = 3
+            kernel_h = 3
+            kernel_w = 3
+            stride_h = 2
+            stride_w = 2
+            pad_h = 1
+            pad_w = 1
+            dilation = 1
+            # Tests the correctness of the conv2d module.
+            in_channels = in_channels_per_group * groups
+            out_channels = out_channels_per_group * groups
+            input_feature_map_size = (H, W)
+            kernel_size = (kernel_h, kernel_w)
+            stride = (stride_h, stride_w)
+            padding = (pad_h, pad_w)
+            dilation = (dilation, dilation)
+            X_scale = 1.3
+            X_zero_point = 2
+            W_scale = [0.5]
+            W_zero_point = [0] if qengine_is_onednn() else [3]
+            Y_scale = 5.0
+            Y_zero_point = 4
+            qconv_cls = nniq.ConvReLU2d
+            module_name = "QuantizedConvReLU2d"
```
nit: can you move these outside of the loop?

leslie-fang-intel replied:
Thanks for the comments. Done.

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -527,26 +633,87 @@ def test_conv2d_api(self):
+        for use_bias, use_channelwise in options:
+            if torch.backends.quantized.engine == "qnnpack":
+                use_channelwise = False
+            batch_size = 2
```
same for these

leslie-fang-intel replied:
Finished.

jerryzh168 commented on test/quantization/core/test_quantized_module.py:
```diff
@@ -597,19 +760,78 @@ def test_conv3d_api(self):
+                [True, False],  # use_channelwise
+            )
+            for pad_mode, use_bias, use_channelwise in options:
+                batch_size = 2
```
same here

leslie-fang-intel replied:
Also finished here.

jerryzh168 commented on torch/ao/nn/intrinsic/modules/fused.py:
```diff
@@ -144,3 +145,13 @@ def __init__(self, linear, tanh):
+class ConvAdd2d(_FusedModule):
+    r"""This is a sequential container which calls the Conv2d modules with extra Add.
+    During quantization this will be replaced with the corresponding fused module."""
+    def __init__(self, add, conv):
```
nit: can we swap the order of conv and add, to be more consistent with the other ops

leslie-fang-intel replied:
Good suggestion, changed the order to make it more consistent.

jerryzh168 reviewed (approved):
looks good, had a few comments inline

jerryzh168 commented on torch/ao/nn/intrinsic/modules/fused.py:
```diff
@@ -144,3 +145,13 @@ def __init__(self, linear, tanh):
+        super().__init__(conv)
+        self.add = add
+
+    def forward(self, x1, x2):
+        return self.add(self[0](x1), x2)
```
also this feels a bit hacky, i'm wondering how we can extend _FusedModule to support this use case..

leslie-fang-intel committed:
Update on "[Quant] Add fused ConvAdd2d module for onednn backend"


**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `ConvAdd2d` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
```
python -m pytest test_quantization.py -k test_conv2d_add
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
